### PR TITLE
common/buffer: add bufferlist::const_iterator

### PR DESCRIPTION
**Pull request by tchaikov**
* unlike its cousin, `bufferlist::const_iterator` is not exported using
  CEPH_BUFFER_API. will do it once we think it will have external users.

Signed-off-by: Kefu Chai <kchai@redhat.com>

### CODE REVIEW

File: src/include/buffer.h [hunk 1]
```diff
@@ -317,6 +318,36 @@ class CEPH_BUFFER_API buffer {
 
     };
 
+    class const_iterator : private iterator {
+    public:
+      const_iterator(const iterator& it)
+	: iterator(it) {}
+      const_iterator(const list *l, unsigned o=0)
+	: iterator(const_cast<list*>(l), o) {}
```
Comment (gregsfortytwo) [inline src/include/buffer.h]: So this is definitely cheating by using a non-const pointer internally. I'm not sure that's a good idea. :/

(You should definitely get other opinions! It just looks to me like asking for trouble.)

Comment (tchaikov) [inline src/include/buffer.h]: yes, generally, we don't define a `const_iterator` using a non-const one. "this approach is sort of controversial", i guessed when i posted this pull request.

an alternative way would be use a template to extract the shared parts of these two iterators, and specialize the different bits. but its readability won't be as good as this one.

i will pull together one later on.

**Comment by tchaikov** [comment]
@gregsfortytwo seems the template exposes the implementation to the user of library and hence requires it to link against libcommon_internal.la which offers the `__ceph_assert_fail()`.

one way to fix this is to use the pimp pattern, but it would lead to unnecessary overhead and complicates the code. so i give up.

**Comment by tchaikov** [comment]
@gregsfortytwo i managed to fix the link problem. could you take a look again? thanks =)

File: src/include/buffer.h [hunk 2]
```diff
@@ -297,11 +309,11 @@ class CEPH_BUFFER_API buffer {
 
       void advance(int o);
       void seek(unsigned o);
-      char operator*();
-      iterator& operator++();
-      ptr get_current_ptr();
+      char operator*() const;
+      iterator_impl& operator++();
+      ptr get_current_ptr() const;
```
Comment (gregsfortytwo) [inline src/include/buffer.h]: So this means that users can't modify individual bytes or buffer pointers any more, right? Does that actually work for us? I thought we overwrote specific bytes in a few places (the ENCODE_START family of macros is what I'm thinking of...but it looks like we use the copy_in function there so I guess it's all good).

**Comment by gregsfortytwo** [comment]
This looks good to me but should get a +1 from @athanatos before merging.

File: src/include/buffer.h (hunk 2 above)
Comment (tchaikov) [inline src/include/buffer.h]: @gregsfortytwo, yes, we do not use 
```c
*bl.begin() = 'c';
```
to modify a byte.

if you think we have future needs for this. i will change the const-ness.

**Comment by tchaikov** [comment]
@gregsfortytwo , thanks for your review.

**Comment by tchaikov** [comment]
@athanatos ping?

**Comment by dachary** [comment]
@tchaikov this needs rebasing.

**Comment by tchaikov** [comment]
`BufferList.rebuild` failed . i am looking at it.

**Comment by liewegas** [comment]
On Thu, 3 Sep 2015, Kefu Chai wrote:
> BufferList.rebuild failed . i am looking at it.

I was just going to say that there are some iterator methods that 
can trigger rebuild.. we need to remove those or something to make a const 
iterator work safely.

**Comment by tchaikov** [comment]
@liewegas thanks for the reminder. yeah, i did remove some of the non-const methods from the const_iterator.

**Comment by tchaikov** [comment]
@athanatos @liewegas i see there are some real-world requirements for the `bufferlist::const_iterator`, for instance:

- https://github.com/ceph/ceph/blob/master/src/test/librbd/test_mock_fixture.h#L39
- https://github.com/ceph/ceph/blob/master/src/messages/MForward.h#L107

so i believe this change will help with the readability and even with the performance sometimes. @gregsfortytwo has helped review an earlier version of this patch. the latest version added the `bool operator!=()` to put the missing piece of the iterator implementation into place.  could you help put another set of eyes on this pull request?

**Comment by liewegas** [comment]
lgtm!

**Comment by tchaikov** [comment]
tested at http://pulpito.ceph.com/kchai-2015-09-15_09:47:40-rados-wip-kefu-testing---basic-multi/

**Pull request merged**